### PR TITLE
maint: Add pre-commit hook to repository

### PR DESCRIPTION
**Pull request by wesbland**

### CODE REVIEW

File: maint/hooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# Checkout a copy of the current index into MIRROR
+git checkout-index --prefix=$MIRROR/ -af
+
+# Remove files from MIRROR which are no longer present in the index
+git diff-index --cached --name-only --diff-filter=D -z HEAD | \
+    (cd $MIRROR && xargs -0 rm -f --)
+
+# This will check the previous commit again when not amending a commit, but that
+# should be ok if the patches are correct.
+filestring=`git diff --cached --name-only --diff-filter=ACM HEAD~1`
```
Comment (wesbland) [inline maint/hooks/pre-commit]: Note that this is a recent change to accommodate amended commits.

It turns out that it's pretty hard to detect whether you're in an amend from inside the hook so I decided to just check the previous commit if creating a new commit. If you're doing it right, it should be safe. If not, just skip the hook.

File: maint/hooks/pre-commit [hunk 2]
```diff
@@ -0,0 +1,94 @@
+#!/bin/sh
```
Comment (raffenet) [inline maint/hooks/pre-commit]: Needs to be `/bin/bash`.

File: maint/hooks/pre-commit [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+IFS=', ' read -r -a file_array <<< "$filestring"
+
+# Everything else happens in the temporary build tree
+pushd $MIRROR > /dev/null
+
+ret=0
+
+for file in "${file_array[@]}"
+do
+    if [[ ($file == *.c || $file == *.h || $file == *.c.in || $file == *.h.in) &&
```
Comment (raffenet) [inline maint/hooks/pre-commit]: We need to make this ignore `src/mpid/ch3/*`.

**Comment by wesbland** [comment]
@raffenet - I've made both of those changes.

**Pull request merged**